Threading attempt

ktf opened the pull request.

ktf committed:
DPL: add optional multithreading to DPL

* Requires updated FairMQ which exposes the ZMQ_EVENTS API.
* Incoming timeslices are put on a queue and processed as tasks by
  a workers pool. By default the pool contains 0 workers and the
  behavior is exactly the same as before.
* Add command line option --threades to specify how many parallel
  threads are used for processing.
* Services of kind Stream are duplicated across workers.
* Message sending contexes are marked as Stream.

ktf commented:
My own work area for multithreaded DPL. Still buggy but the simple tests seem to for a bit until I encounter the error:

```
Could not determine the timeslice for input
```

I am investigating.

bazinski commented on DataFormats/Detectors/TRD/src/RawData.cxx:
```diff
@@ -98,15 +98,15 @@ std::ostream& operator<<(std::ostream& stream, const TrackletMCMData& tracklet)
 {
-  LOGF(INFO, "TrackletMCMData: Raw:0x%08x pos:%d slope:%d pid:0x%08x checkbit:0x%02x",
-       tracklet.word, tracklet.pos, tracklet.slope, tracklet.pid, tracklet.checkbit);
+//  LOGF(INFO, "TrackletMCMData: Raw:0x%08x pos:%d slope:%d pid:0x%08x checkbit:0x%02x",
```
LOGF is not thread safe I take it? I can change this to LOG(info) instead? 
can leave this commented out for now though.

ktf replied:
No, I actually fixed all of them in a different PR. will rebase this one.

ktf committed:
DPL: handle SIGSEGV by dumping the stacktrace and exiting

ktf commented:
@davidrohr as discussed this introduces a `--threads` option, which works like `--pipeline`, i.e.  `--threads <processor-name>:<num-threads>`.

ktf committed:
DPL Foundation: add own runtime_error.

A few features:

* No need to include <stdexcept> hopefully reducing bload.
* No allocation on throw.
* Includes backtrace.

We should use this to remove all the runtime_error from at least the framework.

ktf committed:
DPL: use Framework/RuntimeError helpers throughout the framework

ktf committed:
DPL: add support for checkpoints

ktf committed:
DPL: introduce service for FairMQRegionInfo management

* Moves responsibilty to track and notify FairMQRegionInfo to a DPL
  service from the DataProcessingDevice.
* Fixes (actually impossible?) data race spotted by Thread Sanitizer.

ktf commented:
@knopers8 can you verify this does not have any regression for QC?

knopers8 commented:
> @knopers8 can you verify this does not have any regression for QC?

OK, will do. I came back from holidays today.

knopers8 commented:
I can't compile QC master with this:
```
[  3%] Generating G__QualityControl.cxx, ../lib/G__QualityControl_rdict.pcm, ../lib/libQualityControl.rootmap
In file included from input_line_11:65:
In file included from /home/pkonopka/alice/sw/slc7_x86-64/O2/0_O2_DATAFLOW-1/include/Framework/DataProcessorSpec.h:13:
In file included from /home/pkonopka/alice/sw/slc7_x86-64/O2/0_O2_DATAFLOW-1/include/Framework/AlgorithmSpec.h:13:
In file included from /home/pkonopka/alice/sw/slc7_x86-64/O2/0_O2_DATAFLOW-1/include/Framework/ProcessingContext.h:13:
In file included from /home/pkonopka/alice/sw/slc7_x86-64/O2/0_O2_DATAFLOW-1/include/Framework/InputRecord.h:14:
In file included from /home/pkonopka/alice/sw/slc7_x86-64/O2/0_O2_DATAFLOW-1/include/Framework/DataRefUtils.h:14:
In file included from /home/pkonopka/alice/sw/slc7_x86-64/O2/0_O2_DATAFLOW-1/include/Framework/RootSerializationSupport.h:13:
/home/pkonopka/alice/sw/slc7_x86-64/O2/0_O2_DATAFLOW-1/include/Framework/TMessageSerializer.h:15:10: fatal error: 'Framework/RuntimeError.h' file not found
#include "Framework/RuntimeError.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~
Error: /home/pkonopka/alice/sw/slc7_x86-64/ROOT/v6-20-02-alice7-1/bin/rootcling: compilation failure (/home/pkonopka/alice/sw/BUILD/2c34d715079e6b69d7b85e1b2d3010b7c96b5720/QualityControl/Framework/G__QualityControl9dd1ad6d87_dictUmbrella.h)
Error copying file (if different) from "/home/pkonopka/alice/sw/BUILD/2c34d715079e6b69d7b85e1b2d3010b7c96b5720/QualityControl/Framework/G__QualityControl_rdict.pcm" to "/home/pkonopka/alice/sw/BUILD/2c34d715079e6b69d7b85e1b2d3010b7c96b5720/QualityControl/lib/G__QualityControl_rdict.pcm".
gmake[2]: *** [Framework/G__QualityControl.cxx] Error 1
gmake[1]: *** [Framework/CMakeFiles/QualityControl.dir/all] Error 2
gmake: *** [all] Error 2
```
Any ideas?

ktf committed:
DPL: cleanups to compile with 1.4.22

ktf commented:
Mmm... Weird. Can you explicitly add O2::FrameworkFoundation to the list of dependencies?

knopers8 commented:
Still the same thing. It seems like headers in Foundation aren't exposed to the outside world.

ktf commented:
@aphecetche any idea?

aphecetche commented:
@ktf  will try to see what's going on

@knopers8  I assume the command that is failing is `aliBuild --defaults o2-dataflow build QualityControl` : is that correct ?

knopers8 commented:
Yes, thanks

aphecetche commented:
@knopers8 @ktf The source of that error is that the FrameworkFoundation is promoted to a "real" library (as opposed to a header-only previously), so the include files should be installed. But they are not because the O2AddLibrary cmake function assumes that for target `A` the includes are within `include/A`, and that's not the case here (includes are under `include/Framework` not under `include/FrameworkFoundation`)
 
_One_ way to solve it is to explicitly install those includes

```
diff --git a/Framework/Foundation/CMakeLists.txt b/Framework/Foundation/CMakeLists.txt
index 7d20f5917..cee6665cd 100644
--- a/Framework/Foundation/CMakeLists.txt
+++ b/Framework/Foundation/CMakeLists.txt
@@ -17,6 +17,9 @@ target_compile_definitions(${targetName} PUBLIC -DDPL_ENABLE_BACKTRACE)
 endif()


+install(DIRECTORY ${CMAKE_CURRENT_LIST_DIR}/include/Framework
+        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+
 o2_add_test(test_FunctionalHelpers NAME test_FrameworkFoundation_test_FunctionalHelpers
             COMPONENT_NAME FrameworkFoundation
             SOURCES test/test_FunctionalHelpers.cxx
```

As far as I've seen, it's not the end of the story though, as even with this one fixed, you'll see some more build errors in QualityControl ....

ktf commented:
@aphecetche thanks. Trying it know.
@knopers8 can you give it another try?

knopers8 commented:
I found one problem which was quick to solve, however, I still have some problems with adapting the usage of ServiceRegistry:
```
/home/pkonopka/alice/sw/SOURCES/QualityControl/master/0/Framework/src/TrendingTask.cxx:52:60: error: taking address of temporary [-fpermissive]
   mDatabase = &services.get<repository::DatabaseInterface>();
                                                            ^
/home/pkonopka/alice/sw/SOURCES/QualityControl/master/0/Framework/src/TrendingTask.cxx:52:60: error: cannot convert 'o2::framework::service_ptr<o2::quality_control::repository::DatabaseInterface, o2::framework::NoLocking>*' to 'o2::quality_control::repository::DatabaseInterface*' in assignment
```
I have tried to use the new interface to register the service as `ServiceKind::Stream`, but it seems to have no influence :
```
  mServices.registerService(framework::TypeIdHelpers::uniqueId<DatabaseInterface>(), (void*)mDatabase.get(), framework::ServiceKind::Stream, 0);
// in QualityControl/Framework/src/PostProcessingRunner.cxx
```
Is there something that I am missing?

ktf commented:
Mmm... You cannot cache the mDatabase, since it's per thread. You probably want to use "Global" and capture a reference to it (rather than a pointer) in mDatabase.

ktf committed:
DPL: do proper endOfStream for test_RegionInfoCallbackService

Somehow breaks only during the threading setup

ktf committed:
DPL: protect against empty messages

Looks like we get an extra message when running multithreaded. This
protects against it, but I am not sure how come I get one in the first
place.

knopers8 commented:
Indeed, you are right. I was also missing
```
  constexpr static framework::ServiceKind service_kind = framework::ServiceKind::Global;
```
in DatabaseInterface declaration.
However, now I have problems due to the loss of polymorphism by moving to references :/ (DatabaseInterface is an interface after all). I am not sure yet how to approach it.

knopers8 commented:
@ktf I've prepared the fixes to make it compile and work with this PR and without (see https://github.com/AliceO2Group/QualityControl/pull/477). With this QC seems to be working fine.

davidrohr commented:
@knopers8 : This seems to still fail with qualitycontrol 1.1.2

@ktf : The clang-format and codechecker errors seem to be real:
```
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Framework/Foundation/test/test_RuntimeError.cxx:25:5: error: ignoring return value of function declared with pure attribute [clang-diagnostic-unused-value]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Framework/Foundation/test/test_RuntimeError.cxx:35:5: error: ignoring return value of function declared with pure attribute [clang-diagnostic-unused-value]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Framework/Foundation/test/test_RuntimeError.cxx:47:5: error: ignoring return value of function declared with pure attribute [clang-diagnostic-unused-value]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Utilities/aliceHLTwrapper/src/EventSampler.cxx:118:31: error: prefer a lambda to boost::bind [modernize-avoid-bind]
```

The GPU CI is again failing due to oom killer.

Would be good if we could converge on this finally.

knopers8 commented:
Indeed, weird. I will have a look.

knopers8 commented:
The issues with QC should have been resolved after https://github.com/alisw/alidist/pull/2484 has been merged

davidrohr commented:
@knopers8 ok, cool, thx. @ktf : Could you rebase the PR? Then I'd check again the TPC workflow with the latest version (without --threads) and then we can hopefully finally merge it :)
